### PR TITLE
`ZScaleInterval` parameter name change

### PR DESCRIPTION
**Pull request by tbowers7**
### Description

This pull request addresses non-uniform number-of-samples keywords among the classes in the ``astropy.visualization.interval`` module.

Adds module-standard ``n_samples`` keyword to``ZScaleInterval``, while preserving backward-compatibility with the existing ``nsamples`` keyword.

### Needs

* https://github.com/astropy/astropy/pull/13824

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/visualization/interval.py
Comment (pllim) [inline astropy/visualization/interval.py:231]: Thanks! I think you should look into using this:

https://github.com/astropy/astropy/blob/d4fc9d9b96f0879c363a8183842e5cf6067fae37/astropy/utils/decorators.py#L278

Also subpackage maintainer should contemplate this API change.

**Comment by pllim** [comment]
Maybe some tests need updating to use this new name?

**Comment by tbowers7** [comment]
> Maybe some tests need updating to use this new name?
> 
> Update: Also check the docs for updates too.

No tests in `astropy.visuallization` use the `nsamples` keyword for `ZScaleInterval`.  (No updates needed.)

No doc changes needed other than the [API file](https://docs.astropy.org/en/stable/api/astropy.visualization.ZScaleInterval.html) -- I assume these are regenerated with each release?

File: docs/changes/visualization/13810.api.rst
Comment (eerovaher) [inline docs/changes/visualization/13810.api.rst:2]: The fact that the keyword change was made using `@deprecated_renamed_argument` is an implementation detail that the typical user shouldn't have to care about. What the users are much more likely to be interested in is why the keyword was changed.

Comment (tbowers7) [reply docs/changes/visualization/13810.api.rst]: Added more user-focused verbiage to change-log.

**Comment by pllim** [comment]
Thank you! Would you mind squashing your commits into one commit?

**Review by Cadair — APPROVED** [review]

File: astropy/visualization/interval.py
Comment (pllim) [inline astropy/visualization/interval.py:235]: Do we need something like this for backward compatibility? Though it can be confusing if user modifies `self.nsamples` and not `self.n_samples`. But without this, the code breaks.
```suggestion
        self.n_samples = n_samples
        self.nsamples = self.n_samples
```

Comment (pllim) [reply astropy/visualization/interval.py:235]: It is possible. Maybe also emit a deprecation warning in that setter?

Comment (WilliamJamieson) [reply astropy/visualization/interval.py:235]: Agreed. Should the `getter` also have a deprecation warning as well?

Comment (eerovaher) [reply astropy/visualization/interval.py:235]: You should be using the `deprecated_attribute` decorator here: https://github.com/astropy/astropy/blob/f298b8bc98c44104d183d56bd67f541f7d273f47/astropy/utils/decorators.py#L206

Comment (tbowers7) [reply astropy/visualization/interval.py:235]: The old keyword/attribute are set and got via `@deprecated` `@property` methods now.

Comment (tbowers7) [reply astropy/visualization/interval.py:235]: > You should be using the `deprecated_attribute` decorator here:
> 
> https://github.com/astropy/astropy/blob/f298b8bc98c44104d183d56bd67f541f7d273f47/astropy/utils/decorators.py#L206

I tried using that decorator, but got
```
TypeError: 'property' object is not callable
```
errors... the simple `@deprecated` did not yield such errors.  Suggestions?

Comment (eerovaher) [reply astropy/visualization/interval.py:235]: Yes, but you have reinvented the wheel.

Comment (tbowers7) [reply astropy/visualization/interval.py:235]: I do that a lot.  Suggested code edits are welcome.

Comment (eerovaher) [reply astropy/visualization/interval.py:235]: I shouldn't have called `deprecated_attribute()` a decorator because it isn't. Furthermore, using it to rename an attribute is not as convenient as it should be. I have opened #13824 to make the function simpler to use. It might be a good idea to rebase this pull request once the other one is merged to take advantage of the new functionality.

**Comment by pllim** [comment]
#13824 is merged. You can use it now after you rebase against `upstream/main`. Thanks! 😸

File: astropy/visualization/interval.py
Comment (tbowers7) [reply astropy/visualization/interval.py:235]: Now using ``deprecated_attribute()`` to access the new attribute using the old attribute name.

File: astropy/visualization/interval.py
Comment (eerovaher) [inline astropy/visualization/interval.py:211]: It would be better to use `.. versionchanged` here, see this example: https://github.com/astropy/astropy/blob/7308ca5326caf96f5230b66deecc0a96caa605fa/astropy/io/votable/table.py#L55-L68

Comment (tbowers7) [reply astropy/visualization/interval.py]: Done.

File: astropy/visualization/interval.py
Comment (larrybradley) [inline astropy/visualization/interval.py:214]:
```suggestion
           ``n_samples`` replaces the deprecated ``nsamples`` argument, which will be
           removed in the future.
```

Comment (tbowers7) [reply astropy/visualization/interval.py]: Changed.

**Review by larrybradley — APPROVED** [review]
Thanks, @tbowers7 !

**Pull request merged**
